src/metabase/email.clj

```diff
@@ -95,12 +98,66 @@
                   (assert (#{:tls :ssl :none :starttls} (keyword new-value))))
                 (setting/set-value-of-type! :keyword :email-smtp-security new-value)))
 
+(defsetting email-max-recipients-per-second
+  (deferred-tru "The maximum number of recipients, across emails, that can be sent per second.
+                In case an email has total number of recipients exceeding this limit, the throttling mechanism will be maxed-out but the exception will be silenced.")
+  :export?    true
+  :type       :integer
+  :visibility :settings-manager
+  :audit      :getter)
+
+(defn- make-email-throttler
+  [rate-limit]
+  (throttle/make-throttler
+   :email
+   :attempt-ttl-ms     1000
+   :initial-delay-ms   1000
+   :attempts-threshold rate-limit))
+
+(defonce ^:private email-throttler (when-let [rate-limit (email-max-recipients-per-second)]
+                                     (make-email-throttler rate-limit)))
+
+(defn check-email-throttle
+  "Check if the email throttler is enabled and if so, throttle the email sending based on the total number of recipients.
+
+  If an email has # of recipients greater than the rate limit:
+    a) Skip the throttle check if we haven't breached the rate limit yet.
+    b) If we have breached the rate limit, throttle like a normal email
+
+  We need a) to to avoid creating regression bugs due to the new throttling mechanism."
+  [email]
+  (when email-throttler
+    (when-some [recipients (into #{} (mapcat email) [:to :bcc])]
+      (let [throttle-threshold (.attempts-threshold ^Throttler email-throttler)
+            throttle-check!    (fn [ignore-if-not-breached?]
+                                 ;; check the the first time to see if we have breached the rate limit
+                                 (throttle/check email-throttler true)
+                                 (let [check-the-rest! (fn []
+                                                            ;; dec because we already checked once above
+                                                           (dotimes [_ (dec (count recipients))]
+                                                             (throttle/check email-throttler true)))]
+                                   (if ignore-if-not-breached?
+                                     (do
+                                      (log/warn "Email throttling is enabled and the number of recipients exceeds the rate limit per second. Skip throttling."
+                                                {:email-subject  (:subject email)
+                                                 :recipients     (count recipients)
+                                                 :max-recipients throttle-threshold})
+                                      (u/ignore-exceptions (check-the-rest!)))
+                                     (check-the-rest!))))]
+        (if (> (count recipients) throttle-threshold)
```

Comment:
This is a bit strict I think, `(> (count recipients) (- throttle-threshold attempts-made-so-far))` is fine.

So e.g. we'd let through (1)(2) for threshold of 2